src/cluster/placement/types.go

```diff
@@ -547,6 +548,42 @@ type Service interface {
 	MarkAllShardsAvailable() (Placement, error)
 }
 
+// Operator manipulates Placement's. Each method takes in the current placement and returns a modified
+// copy (without mutating the original placement).
+type Operator interface {
```

Comment:
should service just embed this interface?

Reply:
Hmm I think it's a bit cleaner to leave them separate from the perspective of naming, e.g. you'd end up with:

```
// if we embed

// current method (no interface change)
AddInstances(candidates []Instance) (newPlacement Placement, addedInstances []Instance, err error)

// new method
AddInstancesToPlacement(curPlacement Placement, candidates []Instance) (newPlacement Placement, addedInstances []Instance, err error)
```